src/outputs/restart_hdf5.hpp

```diff
@@ -0,0 +1,339 @@
+using H5D = UnusedPlaceholder;
+using H5S = UnusedPlaceholder;
+#endif
+
+#include "mesh/domain.hpp"
+#include "utils/error_checking.hpp"
+
+namespace parthenon {
+
+class Mesh;
+class Param;
+
+class RestartReaderHDF5 : public RestartReader {
+ public:
+  RestartReaderHDF5(const char *filename);
+
+  [[nodiscard]] SparseInfo GetSparseInfo() const override;
+
+  [[nodiscard]] MeshInfo GetMeshInfo() const override;
+
+  [[nodiscard]] TimeInfo GetTimeInfo() const override;
+
+  [[nodiscard]] std::string GetInputString() const override {
+    return GetAttr<std::string>("Input", "File");
+  };
+
+  // Return output format version number. Return -1 if not existent.
+  [[nodiscard]] int GetOutputFormatVersion() const override;
+
+ private:
+  struct DatasetHandle {
+    hid_t type;
+    H5D dataset;
+    H5S dataspace;
+    int rank;
+    hsize_t count;
+    std::vector<hsize_t> dims;
+  };
+
+  // internal convenience function to open a dataset, perform some checks, and get
+  // dimensions
+  template <typename T>
+  DatasetHandle OpenDataset(const std::string &name) const {
+#ifndef ENABLE_HDF5
+    PARTHENON_FAIL("Restart functionality is not available because HDF5 is disabled");
+#else  // HDF5 enabled
+    DatasetHandle handle;
+
+    // make sure dataset exists
+    auto status = PARTHENON_HDF5_CHECK(H5Oexists_by_name(fh_, name.c_str(), H5P_DEFAULT));
+    PARTHENON_REQUIRE_THROWS(
+        status > 0, "Dataset '" + name + "' does not exist in HDF5 file " + filename_);
+
+    // open dataset
+    handle.dataset = H5D::FromHIDCheck(H5Dopen2(fh_, name.c_str(), H5P_DEFAULT));
+    handle.dataspace = H5S::FromHIDCheck(H5Dget_space(handle.dataset));
+
+    // get the HDF5 type from the template parameter and make sure it matches the dataset
+    // type
+    T *typepointer = nullptr;
+    handle.type = getHDF5Type(typepointer);
+    const H5T dset_type = H5T::FromHIDCheck(H5Dget_type(handle.dataset));
+    status = PARTHENON_HDF5_CHECK(H5Tequal(handle.type, dset_type));
+    PARTHENON_REQUIRE_THROWS(status > 0, "Type mismatch for dataset " + name);
+
+    // get rank and dims
+    const H5S filespace = H5S::FromHIDCheck(H5Dget_space(handle.dataset));
+    handle.rank = PARTHENON_HDF5_CHECK(H5Sget_simple_extent_ndims(filespace));
+
+    handle.dims.resize(handle.rank);
+    PARTHENON_HDF5_CHECK(H5Sget_simple_extent_dims(filespace, handle.dims.data(), NULL));
+    handle.count = 1;
+    for (int idir = 0; idir < handle.rank; idir++) {
+      handle.count = handle.count * handle.dims[idir];
+    }
+
+    return handle;
+#endif // ENABLE_HDF5
+  }
+
+ public:
+  // Gets data for all blocks on current rank.
+  // Assumes blocks are contiguous
+  // fills internal data for given pointer
+  void ReadBlocks(const std::string &name, IndexRange range, std::vector<Real> &dataVec,
+                  const std::vector<size_t> &bsize, int file_output_format_version,
+                  MetadataFlag where, const std::vector<int> &shape = {}) const override {
```

Comment:
Should this be declared here and defined in the cpp file?

Reply:
yes, this was a leftover from the original implementation and I'll clean it up.